### ISSUE TITLE
Bug: [no-inner-declarations] `blockScopedFunctions: "allow"` raises error on strict mode inside function.

### ISSUE THREAD

**Issue by Yukinosuke-Takada** [eslint/eslint#19955]
### Environment

ESLint Playground

### What parser are you using?

Default (Espree)

### Playground Code
```js
/*eslint no-inner-declarations: ["error", "functions", { blockScopedFunctions: "allow" }]*/

// --

function doSomethingElse() {
    "use strict";

    if (test) {
        function doAnotherThing() { }
    }
}

// --
```

### ESLint Configuration 
<details> 
 <summary> Tap to see full config </summary> 

```js
undefined
export default [
    {
        "languageOptions": {
            "sourceType": "script",
            "ecmaVersion": 5
        }
    }
];
```
 </details>

### ESLint Output
```
9:9 Move function declaration to function body root. (no-inner-declarations)
```

### What did you expect to happen?

Please correct me if I'm wrong, isn't `blockScopedFunctions: "allow"` supposed to allow declaring a function if it's strict? The code does not raise an error if its target is set to 2015 or above.

### What actually happened?

9:9 Move function declaration to function body root. (no-inner-declarations)

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8taW5uZXItZGVjbGFyYXRpb25zOiBbXCJlcnJvclwiLCBcImZ1bmN0aW9uc1wiLCB7IGJsb2NrU2NvcGVkRnVuY3Rpb25zOiBcImFsbG93XCIgfV0qL1xuXG4vLyAtLVxuXG5mdW5jdGlvbiBkb1NvbWV0aGluZ0Vsc2UoKSB7XG4gICAgXCJ1c2Ugc3RyaWN0XCI7XG5cbiAgICBpZiAodGVzdCkge1xuICAgICAgICBmdW5jdGlvbiBkb0Fub3RoZXJUaGluZygpIHsgfVxuICAgIH1cbn1cblxuLy8gLS0iLCJvcHRpb25zIjp7InJ1bGVzIjp7fSwibGFuZ3VhZ2VPcHRpb25zIjp7InNvdXJjZVR5cGUiOiJzY3JpcHQiLCJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319LCJlY21hVmVyc2lvbiI6NX19fQ==

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by mdjermanovic** [comment]
Hi @Yukinosuke-Takada, thanks for the issue!

This works as intended, since your config sets `"ecmaVersion": 5`.

https://eslint.org/docs/latest/rules/no-inner-declarations#options

> * `{ blockScopedFunctions: "allow" }` (default) this option allows `function` declarations in nested blocks when code is in strict mode (code with `"use strict"` tag or ESM modules) **and `languageOptions.ecmaVersion` is set to `2015` or above**. This option can be disabled by setting it to `"disallow"`.

In environments that don't support ECMAScript 2015, inner functions are either not block scoped or cause syntax error, so the `blockScopedFunctions` option doesn't apply.

**Comment by Yukinosuke-Takada** [comment]
@mdjermanovic, appreciate the reply. I misunderstood that it had to be strict or ES6+. (I thought it had to satisfy either environment)
